clang/test/dpct/thrust_sort_by_key_usm.cu

```diff
@@ -0,0 +1,79 @@
+// UNSUPPORTED: cuda-8.0
+// UNSUPPORTED: v8.0
+// RUN: dpct -out-root %T/thrust_sort_by_key_usm %s --cuda-include-path="%cuda-path/include" --usm-level=restricted
+// RUN: FileCheck --input-file %T/thrust_sort_by_key_usm/thrust_sort_by_key_usm.dp.cpp --match-full-lines %s
+
+#include <thrust/sort.h>
+#include <thrust/device_vector.h>
+#include <thrust/host_vector.h>
+#include <thrust/execution_policy.h>
+
+int main(void) {
+
+  thrust::device_vector<int> AD(4);
+  thrust::device_vector<int> BD(4);
+  thrust::device_vector<int> SD(4);    
+  thrust::host_vector<int> AH(4);
+  thrust::host_vector<int> BH(4);
+  thrust::host_vector<int> SH(4);  
+  
+  int *h_ptr;
+  int *d_ptr;
+
+  h_ptr = (int*)std::malloc(20 * sizeof(int));
+  cudaMalloc(&d_ptr, 20 * sizeof(int));
+
+/*******************************************************************************************
+ 1. Test sort_by_key
+ 2. Test four VERSIONs (with/without exec argument) AND (with/without comparator argument)
+ 3. Test each VERSION with (device_vector/host_vector/malloc-ed memory/cudaMalloc-ed memory)
+ *******************************************************************************************/
+
+/*********** sort_by_key ***********************************************************************************************************************************************/
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// CHECK:dpct::sort(oneapi::dpl::execution::seq, AH.begin(), AH.end(), BH.begin());
+// CHECK-NEXT:dpct::sort(oneapi::dpl::execution::make_device_policy(q_ct1), AD.begin(), AD.end(), BD.begin());
+// CHECK-NEXT:dpct::sort(oneapi::dpl::execution::seq, h_ptr, h_ptr + 4, BH.begin());
+// CHECK-NEXT:dpct::sort(oneapi::dpl::execution::seq, d_ptr, d_ptr + 4, BD.begin());

```

Comment:
From line43, d_ptr, BD.begin() are device ptr and device iterator,  so the policy should be `oneapi::dpl::execution::make_device_policy(q_ct1)`.

Reply:
SYCLomatic cannot differentiate between h_ptr and d_ptr without data-flow analysis.  The types of h_ptr and d_ptr are both "int *".    The execution policy is chosen by the test:

   CheckThrustArgType(1, "thrust::device_ptr")
   
 The same problem occurs with other functions.  For example:
 
     void foo() {
      thrust::device_vector<int> BD(4);
      int *d_ptr;
      cudaMalloc(&d_ptr, 20 * sizeof(int));
      thrust::unique_by_key(d_ptr,      d_ptr+4,  BD.begin());
    }

Migrates to:

    void foo() {
      dpct::device_vector<int> BD(4);
      int *d_ptr;
      d_ptr = sycl::malloc_device<int>(20, dpct::get_default_queue());
      dpct::unique(oneapi::dpl::execution::seq, d_ptr, d_ptr + 4, BD.begin());
    }

Reply:
For USM case, SYCL provides get_pointer_type() to identify the type of the pointer. We have two options:
1. Generate a if stmt in the migrated code, just like what we did in non-USM;
2. Create a helper function and put the if stmt in the helper function.

Or we can merge the current PR and leave the rest work for future improvement.